### ISSUE TITLE
AES_OFB(data + hash(data)) is insecure, AES-CBC+HMAC or similar must be used

### ISSUE THREAD

**Issue by michwill** [zero-db/zerodb#2]
[Full description](https://www.reddit.com/r/netsec/comments/3vsl6f/zerodb_an_endtoend_encrypted_database_is_open/cxqghim) of vulnerability by [holyvier](https://www.reddit.com/user/holyvier).

[Link to the code](https://github.com/zero-db/zerodb/blob/07df30dc839fc978b07f39b0808f0aa7d5a42ffb/zerodb/transform/encrypt_aes.py#L9)

The way the signature is done isn't standard and it's open to quite a lot of attack. The main problem is that it's not authenticated and the signature is computed before encryption. On top of that, it's using the mode of encryption OFB (Stream cipher mode) which makes altering data easy.

Example :

You have the following data encrypted :
```
data1 = "Username=bob Admin=0"
```

The program computes this :
```
encrypted1 = AES_OFB(data1 + sha256(data1)) + IV1
````

And you want to change it to following :
```
data2 = "Username=bob Admin=1"
```

You simply have to compute this to alter it :
```
encrypted2 = encrypted1 XOR (data1 + sha256(data1)) XOR (data2 + sha256(data2)) + IV1
```

Why has the OFB mode been chosen instead of an authenticated mode like GCM or CBC+HMAC ?

**Comment by thatguystone** [comment]
For the curious (or anyone who stumbles here from reddit), here's a POC:

```python
from Crypto.Cipher import AES
from hashlib import sha256
import os

def sxor(s1,s2):
    return ''.join(chr(ord(a) ^ ord(b)) for a,b in zip(s1,s2))

mode = AES.MODE_OFB
iv_size = 16

iv = os.urandom(iv_size)
key = sha256('test').digest()

pt0 = 'yes yes yes yes '
pt0h = sha256(pt0).digest()

c = AES.new(key, mode, iv)
ct0 = c.encrypt(pt0 + pt0h)

pt1 = 'no no no no no  '
pt1h = sha256(pt1).digest()

c = AES.new(key, mode, iv)
ct1 = c.decrypt(sxor(sxor(ct0, pt0 + pt0h), pt1 + pt1h))

print(ct1[:16])
```

**Comment by michwill** [comment]
All right, so I used MODE_GCM from pycryptodome which we use anyway. We used that particular library because it makes use of AES-NI hardware acceleration. But it appears, it can also has GCM out of the box (which is not present in PyCrypto).

Here's the commit https://github.com/zero-db/zerodb/commit/376055ce20d4738491c75d7209d27107ed6fce93

Thanks again for pointing out this known plaintext attack. Having a working example of it was nice to see :-)
